### PR TITLE
Masking of invalid actions, forward sample, backward sample from model policy and other improvements

### PR DESCRIPTION
**Pull request by alexhernandezgarcia**

### CODE REVIEW

File: aptamers.py [hunk 1]
```diff
@@ -243,7 +243,10 @@ def get_parents(self, state=None, done=None):
             parents = []
             actions = []
             for idx, a in enumerate(self.action_space):
-                if state[-len(a) :] == list(a):
+                is_parent = state[-len(a) :] == list(a)
+                if not isinstance(is_parent, bool):
+                    is_parent = all(is_parent)
```
Comment (manhbao-nguyen) [inline aptamers.py:248]: This is to adapt to the case when the action can be of len > 1, right ? :)

File: gflownet.py [hunk 2]
```diff
@@ -576,49 +647,41 @@ def sample_batch(
...
-                high = (len(self.env.action_space) + 1) * np.ones(len(envs), dtype=int)
-                if self.mask_eos:
-                    high[mask] -= 1
-                actions = self.rng.integers(low=0, high=high, size=len(envs))
+            # Forward sampling
+            if train is False:
+                envs, actions, valids = self.forward_sample(
+                    envs, times, policy="model", model=model, temperature=1.0
+                )
+            else:
```
Comment (manhbao-nguyen) [inline gflownet.py:666]: In the training phase, we  could separate the envs in the envs that correspond to policy = "uniform" and those that correspond to policy = "model"!  For each envs to sample an action for, we can generate a random number to chose if it will chose an action according to the model or uniform and we call separately self.forward_sample with a different policy for the two cases.

File: gflownet.py [hunk 3]
```diff
@@ -487,25 +490,92 @@ def __init__(
...
+            Model to use as policy if policy="model"
+
+        temperature : float
+            Temperature to adjust the logits by logits /= temperature
+        """
+        if not isinstance(envs, list):
+            envs = [envs]
+        states = [env.state2obs() for env in envs]
+        mask_invalid_actions = [env.get_mask_invalid_actions() for env in envs]
+        random_action = self.rng.uniform()
```
Comment (manhbao-nguyen) [inline gflownet.py:519]: random_action is not used later, is it ?

File: gflownet.py [hunk 4]
```diff
@@ -487,25 +490,92 @@ def __init__(
...
+        t0_a_model = time.time()
+        if policy == "model":
+            with torch.no_grad():
```
Comment (manhbao-nguyen) [inline gflownet.py:522]: do we have to specify model.train(mode = False), to remove the dropout option when sampling with the model ?

File: gflownet.py [hunk 5]
```diff
@@ -576,49 +647,41 @@ def sample_batch(
...
                 if valid:
-                    parents, parents_a = env.get_parents(state, env.done)
+                    parents, parents_a = env.get_parents()
                     if train:
                         batch.append(
                             [
                                 tf([env.state2obs()]),
-                                tl(action),
-                                state,
+                                tl([action]),
```
Comment (manhbao-nguyen) [inline gflownet.py:683]: tl([action]) is not used later, is it ?

File: gflownet.py [hunk 6]
```diff
@@ -716,6 +780,7 @@ def flowmatch_loss(self, it, batch):
             with torch.no_grad():
                 next_q = self.target(sp)
         else:
+            # TODO: potentially mask invalid actions next_q
```
Comment (manhbao-nguyen) [inline gflownet.py:783]: I suggest that we directly add the the mask in the training batch information, so that we don't have to compute it another time ! And then it is straightforward to mask the invalid actions in the loss function (if the mask is 1 for valid action and 0 for invalid after_masking_actions = torch.where(masks == 1, next_q, -loginf_tensor))

File: aptamers.py (hunk 1 above)
Comment (alexhernandezgarcia) [reply aptamers.py:248]: Exactly!

File: gflownet.py (hunk 2 above)
Comment (alexhernandezgarcia) [reply gflownet.py:666]: Yes, I had thought of that indeed. The way I would implement it is actually not by separating the envs outside `forward_sample()` but as a new `policy`. Then, inside `forward_sample()` different envs would be sampled differently. However, I believe in practice we will just use `policy="model"` with a `temperature > 1`, since that seems a lot better than uniform sampling.

File: gflownet.py (hunk 3 above)
Comment (alexhernandezgarcia) [reply gflownet.py:519]: True, good catch! I forgot to delete that line. Will do it.

File: gflownet.py (hunk 4 above)
Comment (alexhernandezgarcia) [reply gflownet.py:522]: Uhmm I am not sure if `no_grad()` does that too, but we are not using dropout anyway, do we?

File: gflownet.py (hunk 5 above)
Comment (alexhernandezgarcia) [reply gflownet.py:683]: I believe it is used at least by `unpack_terminal_states()`

File: gflownet.py (hunk 6 above)
Comment (alexhernandezgarcia) [reply gflownet.py:783]: Yes, indeed.

File: gflownet.py (hunk 4 above)
Comment (manhbao-nguyen) [reply gflownet.py:522]: I thought that no_grad() only disabled the heavy computation of gradients. 
Oh yes, I mixed up with another MLP that I wrote in the same spirit as the current proxy for the GFN, and where I added dropout for the training. :)
Current implementation is indeed  calling Sequential built-in MLP !
